server/src/main/java/org/elasticsearch/rest/action/search/HttpChannelTaskHandler.java

```diff
@@ -0,0 +1,158 @@
+                    //the flag, we would just miss unlinking for some tasks that are being linked while onResponse is called.
+                    synchronized(task) {
+                        try {
+                            //nothing to do if link was not called yet: we would not find the task anyways.
+                            if (link.getAndIncrement() > 0) {
+                                CloseListener closeListener = httpChannels.get(httpChannel);
+                                TaskId taskId = new TaskId(client.getLocalNodeId(), task.getId());
+                                closeListener.unregisterTask(taskId);
+                            }
+                        } catch(Exception e) {
+                            listener.onFailure(e);
+                        }
+                    }
+                }
+            });
+
+        CloseListener closeListener = httpChannels.computeIfAbsent(httpChannel, channel -> new CloseListener(client));
+        synchronized (task) {
+            //the task will only be registered if it's not completed yet, meaning if its TaskListener has not been called yet.
+            //otherwise, given that its listener has already been called, the task id would never be removed.
+            if (link.getAndIncrement() == 0) {
+                TaskId taskId = new TaskId(client.getLocalNodeId(), task.getId());
+                closeListener.registerTask(httpChannel, taskId);
+            }
+        }
+
+        //TODO test case where listener is registered, but no tasks have been added yet:
+        // - connection gets closed, channel will be removed, no tasks will be cancelled
+
+        //TODO check that no tasks are left behind through assertions at node close
+    }
+
+    final class CloseListener implements ActionListener<Void> {
+        private final Client client;
+        final Set<TaskId> taskIds = new CopyOnWriteArraySet<>();
+        private final AtomicReference<HttpChannel> channel = new AtomicReference<>();
+
+        CloseListener(Client client) {
+            this.client = client;
+        }
+
+        void registerTask(HttpChannel httpChannel, TaskId taskId) {
+            if (channel.compareAndSet(null, httpChannel)) {
+                //In case the channel is already closed when we register the listener, the listener will be immediately executed which will
+                //remove the channel from the map straight-away. That is why we do this in two stages. If we provided the channel at close
+                //listener initialization we would have to deal with close listeners calls before the channel is in the map.
+                httpChannel.addCloseListener(this);
+            }
+            this.taskIds.add(taskId);
+        }
+
+        private void unregisterTask(TaskId taskId) {
+            this.taskIds.remove(taskId);
+        }
+
+        @Override
+        public void onResponse(Void aVoid) {
```

Comment:
I'd also synchronize this call in order to ensure that it cannot execute concurrently.